psi4/src/psi4/libqt/timer.cc

```diff
@@ -66,243 +67,997 @@
 ** J. F. Gonthier, February 2016
 */
 
+#include "psi4/libciomr/libciomr.h"
+#include "psi4/libparallel/ParallelPrinter.h"
+#include "psi4/psi4-dec.h"
+#include "psi4/psifiles.h"
 #include <cstdio>
 #include <cstdlib>
-#include <unistd.h>
 #include <cstring>
 #include <ctime>
 #include <sys/param.h>
 #include <sys/times.h>
-#include "psi4/libciomr/libciomr.h"
-#include "psi4/psifiles.h"
-#include "psi4/psi4-dec.h"
-#include "psi4/libparallel/ParallelPrinter.h"
+#include <unistd.h>
+
+#include <chrono>
+#include <list>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+#ifdef _OPENMP
+#include <omp.h>
+#else
+#define omp_get_thread_num() 0
+typedef int omp_lock_t;
+#define omp_init_lock(lock_timer_p)                                            \
+  do {                                                                         \
+  } while (0)
+#define omp_set_lock(lock_timer_p)                                             \
+  do {                                                                         \
+  } while (0)
+#define omp_unset_lock(lock_timer_p)                                           \
+  do {                                                                         \
+  } while (0)
+#define omp_destroy_lock(lock_timer_p)                                         \
+  do {                                                                         \
+  } while (0)
+#endif
+
 /* guess for HZ, if missing */
 #ifndef HZ
 #define HZ 60
 #endif
 
-#define TIMER_KEYLEN 128
-#define TIMER_OFF 0
-#define TIMER_ON 1
-
 namespace psi {
 
-struct timer {
-    char key[TIMER_KEYLEN];
-    unsigned int status;
-    unsigned int calls;
-    double utime;
-    double stime;
-    double wtime;
-    struct tms ontime;
-    struct timeval wall_start;
-    struct timer *next;
-    struct timer *last;
-};
-
-struct timer *global_timer;
-time_t timer_start, timer_end;  /* Global wall-clock on and off times */
+using clock = std::chrono::high_resolution_clock;
 
-/*!
-** timer_init(): Initialize the linked list of timers
-**
-** \ingroup QT
-*/
-void timer_init(void)
-{
-  extern struct timer *global_timer;
-  extern time_t timer_start;
+enum Timer_Status { OFF, ON, PARALLEL };
 
-  timer_start = time(NULL);
+class Timer_Structure;
 
-  global_timer = NULL;
-}
+class Timer_thread {
+private:
+  Timer_Status status_;
+  size_t n_calls_;
+  clock::time_point wall_start_;
+  clock::duration wtime_;
 
-/*!
-** timer_done(): Close down all timers and write results to timer.dat
-**
-** \ingroup QT
-*/
-void timer_done(void)
-{
-
-  char *host;
-  extern time_t timer_start, timer_end;
-  struct timer *this_timer, *next_timer;
-  extern struct timer *global_timer;
-
-  timer_end = time(NULL);
-
-  host = (char *) malloc(40 * sizeof(char));
-  gethostname(host, 40);
+public:
```

Comment:
Can you reformat this with 4-space indentions? Simplest way would be to use `clang-format` with google style options (4-spaces).